include/cuco/bloom_filter.cuh

```diff
@@ -0,0 +1,393 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/bloom_filter_ref.cuh>
+#include <cuco/detail/storage/storage_base.cuh>
+#include <cuco/extent.cuh>
+#include <cuco/hash_functions.cuh>
+#include <cuco/utility/allocator.hpp>
+#include <cuco/utility/cuda_thread_scope.cuh>
+
+#include <cuda/atomic>
+#include <cuda/std/array>
+#include <cuda/stream_ref>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+
+namespace cuco {
+
+/**
+ * @brief A GPU-accelerated Blocked Bloom filter.
+ *
+ * The `bloom_filter` supports two types of operations:
+ * - Host-side "bulk" operations
+ * - Device-side "singular" operations
+ *
+ * The host-side bulk operations include `add`, `test`, etc. These APIs should be used when
+ * there are a large number of keys to add or lookup. For example, given a range of keys
+ * specified by device-accessible iterators, the bulk `add` function will add all keys into
+ * the filter.
+ *
+ * The singular device-side operations allow individual threads (or Cooperative Groups) to perform
+ * independent add or lookup operations from device code. These operations are accessed through
+ * non-owning, trivially copyable reference types (or "ref").
+ *
+ * @note `Block` is used **only** to determine `block_words` via `cuda::std::tuple_size<Block>` and
+ * `word_type` via `Block::value_type` and does not represent the actual storage type of the filter.
+ * We recommend using `cuda::std::array`.
+ *
+ * @tparam Key Key type
+ * @tparam Block Type to determine the filter's block size and underlying word type
+ * @tparam Extent Size type that is used to determine the number of blocks in the filter
+ * @tparam Scope The scope in which operations will be performed by individual threads
+ * @tparam Hash Hash function used to generate a key's fingerprint
+ * @tparam Allocator Type of allocator used for device-accessible storage
+ */
+template <class Key,
+          class Block              = cuda::std::array<std::uint64_t, 4>,
```

Comment:
I'd recommend `<std::uint32_t, 8>` instead.

First, there's what Felipe said about the intersection false positive probability, which improves a lot for partitioned/split/segmented filters - so setting 8 bits in four 64-bit words doesn't have nearly as nice characteristics as setting 8 bits in eight 32-bit words.

OK, you might say, but why set 8 bits at all - isn't four bits a perfectly good default? Well, this produces (depending on filter size vs. population and increasing because of the blocking) a false-positive probability of 6.25%. This is high enough that I don't think I've seen it used in any research on the topic. 8 bits is 0.4% - which is definitely in the range of what is used in state-of-the-art applications in databases and networking.

Third, `<std::uint64_t, 8>` doesn't improve the false-positive probability much over `<std::uint32_t, 8>` - you can do the math or run experiments to demonstrate as such. See `libfilter_block_fpp_detail` in https://github.com/jbapple/libfilter/blob/4ebeaef1215969aee9edb05eb145e94b8dd98e16/c/lib/util.c#L5, where `ndv` is the *n*umber of *d*istinct *v*alues, `bytes` is the total space used by the filter, `word_bits` is 32 vs. 64, `bucket_words` is 8, and `hash_bits` is 64 or 32 (shouldn't change the false-positive probability by more than a tiny, tiny amount). The calculation is derived from from "Cache-, Hash- and Space-Efficient Bloom Filters", by FELIX PUTZE, PETER SANDERS and JOHANNES SINGLER, equation (3).

Finally, there's the hope of direct compatibility with the wealth of projects using `<std::uint32_t, 8>`. Perhaps cuCollections could even copy a filter directly from persistent storage or host memory and use it directly - without needing to build it from the raw keys! This could enable very large speedups for certain applications.

Reply:
Thank you for the detailed explanation. Very convincing. There wasn't much difference between 8*4byte vs. 4*8byte setup in the benchmarks so I just picked one of them at random. Still it seems the 1*4byte is faster but the FPR drops significantly. So it's a trade-of in the end. I'd say 8*4byte is a good sweet spot. It's still a tuneable parameter so users can adjust it towards their needs.

> The calculation is derived from from "Cache-, Hash- and Space-Efficient Bloom Filters", by FELIX PUTZE, PETER SANDERS and JOHANNES SINGLER, equation (3).

I think I read that paper at least 10 times :D It would also be helpful for our case to provide some helper function to determine the optimal filter parameters. This could be a separate PR.

Reply:
If you look at the link I posted, you'll see two other functions that support  calculating the minimal space needed for a given NDV and FPP and calculating the maximum NDV supported given the FPP needed and the amount of space provided. It's licensed Apache 2.0

Reply:
Awesome! Also, great that it has a friendly license ;) I'd have to start an internal SWIPAT process for legal reasons so I can adopt your implementation or ideas from it. Of course this would be attributed in the copyright header. Are you cool with that?

Reply:
Sure thing 🙂